### DISCUSSION TITLE
Code coverage for the .NET application using robot framework automation test cases

### DISCUSSION THREAD

**Discussion by RamyaKT** [SteveGilham/altcover#210] [Q&A]
Hi Team,

I have the source code in .NET application. but for automation, I have written test cases in robot framework (on top of python). From the QA perspective, I wanted to get the code coverage through running automation test cases.

is there any way to achieve this using altCover?

I am not finding the relevant documents / details regarding cross language platform. could anyone please help me out for this.
If anyone having the examples on connecting 2 different platforms would also be very helpful.

**Comment by SteveGilham** [comment]
AltCover only works for code using .Net (framework or cross-platform, including Mono), as it works by rewriting the compiled IL in the resulting assemblies. For interpreted python code, you need a tool that works with the tracing hooks in that runtime; the one I am familiar with there is [`coverage.py`](https://pypi.org/project/coverage/).

**Reply by SteveGilham** [reply]
For the .Net portion of your code, so long as those assemblies have been instrumented, then however they are executed, they will make the calls via the recorder assembly that log the code that has been visited. The process that would be followed is to deploy the code to the test location, instrument it, run the test suite over the instrumented code, then gather the coverage data from the test, following the steps noted [here](https://github.com/SteveGilham/altcover/wiki/Modes-of-Operation#instrument-now-test-later-collect-coverage-after-that).

For the python portion, to get a metric of the coverage of the test code, then the tests of the instrumented code should be run under `coverage.py` or equivalent.

**Reply by RamyaKT** [reply]
Hi Steve,

I was trying to add the nuget package of altcover. after adding I build the source project & intrumented it. while running the runner, I got the error as "Recorder assembly E:\xxx\AltCover.Recorder.g.dll not found in the nominated directory."

When instrumented I got the "AltCover.Monitor.dll" but not the recorder.

Could you please help me out to get the recorder dll?

**Reply by RamyaKT** [reply]
Currently the source project is in .NET 6.0. will this altcover will work for this project?

**Reply by SteveGilham** [reply]
Taking the easy question first, yes, anything .Net Framework 2.0 or later is supported, up to and including net9.0pv.

About the recorder assembly - that is written as the last step of instrumentation, along with all the ; if you use `--inplace`, it should be in the input directory, otherwise it should be in the output directory (defaulting to `./__Instrumented`) along with the other instrumented assemblies.  You can quickly check an assembly has been instrumented by opening the manifest in `ILDasm.exe`

![Screenshot 2024-03-05 133205](https://github.com/SteveGilham/altcover/assets/7460854/aa9b50cf-d1d1-4d18-bbb5-99feac80dd57)

as it will have a reference to the recorder and include an assembly-level `[Instrumentation]` attribute.  The test should, of course, be run on the instrumented copies of the application.  The collection phase of the process looks to find the recorder assembly, as that contains information about the coverage report file (location, report format, etc) from the instrumentation process, to ensure consistency.

**Reply by RamyaKT** [reply]
Hi Steve,

1. I was able to instrument it properly. I used the code as "altcover" in cmd which created me a folder called "__Instrumented" inside which I am able to all required dlls of my project & also of altcover. I hope my instrumentation method is correct.
2. After instrumenting I needed to run the application to run the APIs. so I gave the command like: " AltCover Runner -r "E:\xxx\__Instrumented" -x "E:\xxx\__Instrumented\yyy.WebService.exe" ". This started my application. through postman I was able to hit the API for the application
4. after this API execution I entered, "CTRL+C" to terminate. and I ran the command "AltCover Runner --collect -r "E:\xxx\__Instrumented". I got the error as below:
    **ERROR *** Collection phase failed.
      The process cannot access the file 'E:\xxx\coverage.xml.0.acv' because it is being used by another process.
      Details written to E:\xxx\__Instrumented\__Instrumented\AltCover-2024-03-05--21-00-01.log**
5. so I had to kill the exe service.
6. after this again I ran the command: AltCover Runner --collect -r "E:\xxx\__Instrumented"
7. I got the response as below:
    ... E:\xxx\coverage.xml.0.acv (86,832b)
    8,830 visits recorded in 00:00:00.4767464 (18,521 visits/sec)
    A total of 8,830 visits recorded
    Coverage statistics flushing took 0.01 seconds
8. but I am not able to find the results xml file.
9. if I give with -o (outputdirectory), xml file will generate but there will be no data in that.

could you please help me to sort it out

**Reply by RamyaKT** [reply]
Hi Steve,

Any suggestions on above concern

**Reply by SteveGilham** [reply]
The tacit assumption is that the process from which coverage is being collected has terminated before coverage processing begins, as that is the only point at which it can be certain that all data have been collected.

The coverage data files - the `...[number].acv` files - remain in use until the `ProcessExit` handler executes, as only then is it obviously safe for the recorder to flush all I/O to file and close the streams without data loss.

The coverage report file indicated by `-r` or, `--report` defaults to `coverage.xml` or `coverage.json` in the current directory where the instrumentation was performed; it is created at instrumentation time as a skeleton with zero coverage values, which the end-of-test processing then simply has to fill in, and the path to it is stored in the generated recorder assembly.  

When the instrumented code executes, the recorder assembly looks at the `[absolute path to coverage reportfile]` which has been written into it, and appends `.acv`, to find the control file (created when `--save` is given, but also susceptible to [manual control](https://github.com/SteveGilham/altcover/wiki/Pause%E2%95%B1Resume-and-the-Control-File)). If it exists, the recorder will create one or more `[absolute path to coverage reportfile].[number].acv` files containing a compressed record of the branch and sequence point visits. If not, the visit data will be stored in memory.

When the process running the instrumented code terminates, if the control file never existed during the run, the `ProcessExit` handler reads the (NCover or OpenCover format) empty coverage report `[absolute path to coverage reportfile]` and fills in with the branch and sequence point visit counts from memory (ignoring all the derived quantities in the OpenCover format). If the control file existed at any point, then any remaining data in I/O buffers is simply flushed to disk at this point.

The function of `altcover runner --collect` is to read `[absolute path to coverage reportfile]` from the recorder, then read all the the `[absolute path to coverage reportfile].[number].acv` files and process them into the original report, which bypasses any process exit time constraints, and handles more elaborate output formats than the basic XML ones.

**Reply by SteveGilham** [reply]
I'm assuming that the recent comment that was made about finding the coverage report and I received an e-mail notification of has been deleted, because you managed to resolve the issue yourself.

Just in case, from the [usage guide](https://stevegilham.com/altcover/Usage) 
```
  -r, --report=VALUE         Optional: The output report template file (default:
                                'coverage.xml' or 'coverage.json' in the
                               current directory)
      --reportFormat=VALUE   Optional: Generate the report in the specified
                               format (Json, NCover or the default OpenCover)
```
to make the location and format of report explicit
